[add] cms_form + cms_form_example

simahawk opened the pull request:
CMS Form
========

Basic website contents form framework. Allows to define front-end forms for every models in a simple way.

If you are tired of re-defining every time an edit form or a search form for your odoo website,
this is the module you are looking for.

Features
========

* automatic form generation (create, write, search)
* automatic route generation (create, write, search)
* automatic machinery based on fields' type:
    * widget rendering
    * field value load (from existing instance or from request)
    * field value extraction (from request)
    * field value write (to existing instance)

* highly customizable
* works with every odoo model
* works also without any model
* add handy attributes to models inheriting from ``website.published.mixin``:
    * ``cms_add_url``: lead to create form view. By default ``/cms/form/create/my.model``
    * ``cms_edit_url``: lead to edit form view. By default ``/cms/form/edit/my.model/model_id``
    * ``cms_search_url``: lead to search form view. By default ``/cms/form/search/my.model``

(extracted from [official README](https://github.com/simahawk/website-cms/blob/add-cms_form-PR/cms_form/README.rst))

simahawk committed:
[add] cms_form + cms_form_example

simahawk commented:
@lasley @Yajo as usual, when you have time.... :)

PS: is it possible to configure test coverage checks to ignore some stuff, like cms_form_example module?

lasley commented:
> PS: is it possible to configure test coverage checks to ignore some stuff, like cms_form_example module?

[Few different ways](https://coverage.readthedocs.io/en/coverage-4.2/excluding.html) - in your instance prolly the coveragerc file would be best.

My warning is to not go excluding things all over the place because it's really just covering up a problem. Anything not tested is a ticking timebomb IMO.

That said, we're definitely fine ignoring an example module.

lasley commented on .coveragerc:
```diff
@@ -0,0 +1,4 @@
+[run]
```
Take a look at https://github.com/LasLabs/python-cfssl/blob/master/.coveragerc

You want report section

simahawk replied:
well, the docs do not mention [report] here https://coverage.readthedocs.io/en/coverage-4.2/source.html#execution
anyway, I updated it as per your example, but is still looking for cms_form_example.
Better look at this tomorrow, instead of checking it in the middle of my baby's bath :smile:

lasley replied:
I looked further and it looks like [coveragerc isn't an option because the template is being used instead](https://github.com/OCA/maintainer-quality-tools/blob/master/travis/travis_install_nightly#L69). It would probably be best to just use
```
# pragma: no cover
```

near the top of the python files then 😦

lasley commented:
@tedsalmon please review

tedsalmon commented on cms_form/models/cms_form.py:
```diff
@@ -0,0 +1,225 @@
+# Copyright 2017 Simone Orsi
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models
```
A little nit picky but these imports are not ordered alphabetically [as OCA guidelines ask](https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#imports). Perhaps you'd consider importing in a single line?

`from  openerp import _, exceptions, models`

tedsalmon commented on cms_form/models/cms_form_mixin.py:
```diff
@@ -0,0 +1,394 @@
+from ..utils import DEFAULT_LOADERS, DEFAULT_EXTRACTORS
+from ..import widgets
+
+import inspect
```
These imports should be above the Odoo import

tedsalmon commented on cms_form/models/cms_form_mixin.py:
```diff
@@ -0,0 +1,394 @@
+                value = extractor(self, fname, value, **request_values)
+            if value is None:
+                # we assume we do not want to override the field value.
+                # a tipical example is an image field.
```
`s/tipical/typical`

tedsalmon commented on cms_form/models/cms_search_form.py:
```diff
@@ -0,0 +1,129 @@
+    def form_search_domain(self, search_values):
+        """Build search domain.
+
+        TODO...
```
What's TODO here?

simahawk replied:
well: explain how we build domain and maybe improve it a little bit.

tedsalmon commented on cms_form/models/test_models.py:
```diff
@@ -0,0 +1,77 @@
+# Copyright 2017 Simone Orsi
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import models
```
Alphabetical import order, again :)

tedsalmon commented on cms_form/tests/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from . import test_form_search
+from . import test_form_render
+from . import test_loaders
+from . import test_controllers
```
Import alphabetically :)

tedsalmon commented on cms_form/tests/test_controllers.py:
```diff
@@ -0,0 +1,91 @@
+from .common import fake_request
+from .common import FormHttpTestCase
+from ..controllers import main
+import mock
```
These should be above your local imports

simahawk replied:
@tedsalmon tnx! I'll go trough them.

tedsalmon reviewed (approved):
👍  Awesome work! I have some minor review items, mostly import related stuff that caught my eye :)

guewen commented on cms_form/README.rst:
```diff
@@ -0,0 +1,182 @@
+            if extra_values.get('notify_partner'):
+                # do what you want here...
+
+``notifiy_partner`` will be included into the form but it will be discarded on create and write.
```
`s/notifiy/notify/g`

guewen reviewed (approved):
Wow

lasley commented:
@simahawk - Did you have any plans to up the test coverage?

simahawk committed:
can_edit: check for access rules too

simahawk commented:
@lasley I hope we can do this https://github.com/OCA/maintainer-quality-tools/issues/435. 
If not, I'm going to hack it w/ `pragma` stuff where is needed.

simahawk committed:
[imp] support selection field

lasley commented:
@simahawk - your PR was merged, so I restarted the Travis build. This should update the coverage!

lasley commented on .coveragerc:
```diff
@@ -0,0 +1,24 @@
+[run]
```
We can remove this file now that we've confirmed MQT doesnt adhere to it

simahawk committed:
remove useless coverage conf

simahawk committed:
[fix] boolean widget value handler

simahawk committed:
add form_to_bool util [ci skip]

lasley commented:
@simahawk - any chance to get this amazingness licensed under LGPL? Not a blocker or anything, just a request.

simahawk commented:
@lasley no strong opinion against it. I'll think about it ;)

lasley commented:
Thanks @simahawk. To elaborate on my reasoning - any form created with this module will need to be AGPL'd, which seriously hinders usage in Enterprise environments IMO

lasley commented:
PS - are you planning on making a form builder of any sort out of this (to rival that of Odoo EE)? I've been tossing the idea around in my head, just want to make sure we don't overlap any effort.

simahawk commented:
@lasley not yet. ATM I'd like to add AJAX submit, then maybe JSON serialization to allow any sort of usage via JS.
Before this I have to move widgets to models and support subobjects creation (o2m/m2m). 
Maybe we can have an HG and talk about all of this? ;)

lasley commented:
> Maybe we can have an HG and talk about all of this? ;)

Sure dave@dlasley.net is my Google. I work almost endlessly, so whenever works for the most part 😛

lasley commented:
For form submission via AJAX, something simple [like this](https://github.com/LasLabs/stock-logistics-barcode/blob/3a098684024cd546da5539de41275c9786b8c396/web_stock/static/src/js/mixins.js#L10) could work?

simahawk commented:
~~hmm, you are binding to the "click" on the button I thought about that but I feel a bit "dirty" in my case :stuck_out_tongue_closed_eyes: I'll try that too anyway.~~
Ouch, sorry, I thought you were referring to the issue in cms_delete_content PR #6  :sweat:

simahawk committed:
add master/slave field support

simahawk committed:
allow override of redirect url via "redirect" param [ci skip]

simahawk committed:
fixup! allow override of redirect url via "redirect" param [ci skip]

simahawk committed:
update german transl [ci skip]

simahawk committed:
fix translation [ci skip]

simahawk committed:
[imp] wrap help block into div to allow custom <p />

simahawk committed:
[fix] request value retrieval w/ extra querystring args

Make sure to give precedence to `request.form` attribute
since you might get some extra params (like ?redirect)
and this will make the form machinery miss all the fields.

This was causing image widgets `keep flags` disappear
and make the form flush the image value.

simahawk committed:
[fix] master/slave: get only changed input

simahawk committed:
[add] selection radio widget

simahawk committed:
add option item class

simahawk committed:
add subwidgets facilities and help msg by option

simahawk committed:
improve master slave js

lasley commented:
@simahawk - I set this to WIP because commits are still being added. Let us know when it's ready for review again.

simahawk commented:
@lasley yes pls :) I'm still cleaning/fixing/adding small improvements and I'll move widgets to models.
Hopefully we can close this during or after the sprint in Barcelona (I will attend remotely).

simahawk committed:
improve search form template

simahawk committed:
add TODO to improve validation [ci skip]

simahawk committed:
[imp] include form name into wrapper css class

simahawk committed:
fix tests

simahawk committed:
[imp] cache form_fields

simahawk committed:
[imp] move widgets to models

simahawk committed:
ease widget override

simahawk committed:
widgets fix/imp

simahawk committed:
better subfields

simahawk committed:
Relicense under LGPL

simahawk commented:
@lasley I'd say that after this https://github.com/OCA/website-cms/pull/9/commits/6f7a5946869f613d979bf6e287da9c6401fbdab6 we can rebase and merge :)
About coverage: I'd like to improve it during the sprint next week but to ease partecipation is better to merge this before so that we already have a "kinda stable" starting point.

simahawk commented:
BTW I'll re-license already merged modules in a separated PR

simahawk committed:
update license in README [ci skip]

simahawk committed:
[fix] load defaults via defaults_get

simahawk committed:
oops, remove print statement

simahawk commented:
@pedrobaeza would you mind hit the button here? I'll improve test cov later, probably tomorrow.
Needs rebase of course...
